### PR TITLE
interface: keep the viewcode in the urls

### PR DESCRIPTION
**Pull request by Garfield-fr**
The change of the user's data as well as the change of
the password has been transferred to the Angular user profile.

* Adds functionality for editing user data in angular.
* Adds functionality to change user password in angular.
* Closes https://github.com/rero/rero-ils/issues/2000.
* Closes https://github.com/rero/rero-ils/issues/2101.
* Closes https://github.com/rero/rero-ils/issues/2195.

Co-Authored-by: Bertrand Zuchuat <bertrand.zuchuat@rero.ch>

## Dependencies

My PR depends on the following PR(s):

* [patron: select the current profile view](https://github.com/rero/rero-ils-ui/pull/801)
* [users: make user profiles read-only](https://github.com/rero/invenio-userprofiles/pull/6)

## How to test?

- Test the data editing functionality in the "Personal data" tab.
- Test the password change functionality in the "Personal data" tab.
- Test the viewcode on Ill request.

## Code review check list

- [ ] Commit message template compliance.
- [ ] Commit message without typos.
- [ ] File names.
- [ ] Functions names.
- [ ] Functions docstrings.
- [ ] Unnecessary commited files?
- [ ] Cypress tests successful?

### CODE REVIEW

File: rero_ils/modules/ill_requests/views.py [hunk 1]
```diff
@@ -35,16 +35,16 @@
...
     __name__,
-    url_prefix='/ill_requests',
+    url_prefix='/<string:viewcode>',
     template_folder='templates',
     static_folder='static',
 )
 
 
-@blueprint.route('/create/', methods=['GET', 'POST'])
+@blueprint.route('/ill_requests/create/', methods=['GET', 'POST'])
```
Comment (zannkukai) [inline rero_ils/modules/ill_requests/views.py:44]: The route `<string:viewcode>/ill_requests/create` seems very strange REST url.
Maybe better :
  * `ill_requests/create?viewcode=xxxx`
  * `ill_requests/<viewcode>/create``

I suggest the first one (seems better REST route). And test if `request.args.get('viewcode')` exists into the method (or extract this argument into a decorator)

File: rero_ils/modules/users/views.py [hunk 2]
```diff
@@ -42,6 +44,24 @@ def is_logged_librarian(*args, **kwargs):
...
+        """Decorated view."""
+        if not current_user.is_authenticated:
+            abort(401)
+        if not current_librarian and 'id' in kwargs \
+                and current_user.id != int(kwargs.get('id')):
+            abort(401)
+        return fn(*args, **kwargs)
+    return is_logged
+
+
```
Comment (zannkukai) [inline rero_ils/modules/users/views.py:64]: NOTE: It should be possible to work with Invenio `Permission`. Maybe to implement after "permission spike US"

File: rero_ils/modules/users/views.py [hunk 3]
```diff
@@ -131,7 +151,11 @@ def get(self):
             user = User.get_by_username_or_email(email_or_username)
         if not user:
             return hits
-        data = user.dumps()
+        # if librarian: send all user data
+        # if patron: send only the user id
+        user_data = user.dumps()
+        data = user_data if current_librarian else \
+            {'id': user_data.get('id')}
```
Comment (zannkukai) [inline rero_ils/modules/users/views.py:158]: Dumping a resource is could be "huge" process. Maybe better in reverse way
```python
data = user.dumps() if current_librarian else {'id': user.id}
```

File: rero_ils/theme/menus.py [hunk 4]
```diff
@@ -128,6 +128,10 @@ def init_menu_tools():
     rero_register(
         item,
         endpoint='ill_requests.ill_request_form',
+        endpoint_arguments_constructor=lambda: dict(
+            viewcode=request.view_args.get(
+                'viewcode', current_app.config.get(
+                    'RERO_ILS_SEARCH_GLOBAL_VIEW_CODE'))),
```
Comment (zannkukai) [inline rero_ils/theme/menus.py:134]: related to my comment about REST endpoint

File: tests/api/users/test_users_rest.py [hunk 5]
```diff
@@ -139,7 +139,7 @@ def test_users_search_api(client, librarian_martigny, patron_martigny):
     assert hits['hits']['hits'] == []
     assert hits['hits']['total']['value'] == 0
 
-    # all by username
+    # # all by username
```
Comment (zannkukai) [inline tests/api/users/test_users_rest.py:142]: remove unnecessary #

File: tests/api/users/test_users_rest.py [hunk 6]
```diff
@@ -148,10 +148,11 @@ def test_users_search_api(client, librarian_martigny, patron_martigny):
     )
     assert res.status_code == 200
     hits = get_json(res)
-    assert hits['hits']['hits'][0]['id'] == patron_martigny['user_id']
+    assert hits['hits']['hits'][0]['metadata']['username'] == \
+        patron_martigny['username']
     assert hits['hits']['total']['value'] == 1
 
-    # all by email
+    # # all by email
```
Comment (zannkukai) [inline tests/api/users/test_users_rest.py:155]: same

File: rero_ils/modules/ill_requests/views.py (hunk 1 above)
Comment (Garfield-fr) [reply rero_ils/modules/ill_requests/views.py:44]: The structure of the url on the site always has as 2nd parameter in the url, the view code.

**Comment by ManaDeweerdt** [comment]
- [ ] We cannot save anymore the personal data. The button is greyed. Ex Katie: https://ilsdev.test.rero.ch/global/patrons/profile

File: rero_ils/modules/users/views.py [hunk 7]
```diff
@@ -42,6 +44,40 @@ def is_logged_librarian(*args, **kwargs):
...
+    The access is allow when the connected user is a librarian or
+    the user id is the same of the id argument.
+    """
+    @wraps(fn)
+    def is_logged(*args, **kwargs):
+        """Decorated view."""
+        if not current_user.is_authenticated:
+            abort(401)
+        if not current_librarian and 'id' in kwargs \
+                and current_user.id != int(kwargs.get('id')):
```
Comment (zannkukai) [inline rero_ils/modules/users/views.py:59]: More readable : `if not current_librarian and str(current_user.id) != kwargs.get('id', None): `

File: rero_ils/theme/menus.py [hunk 8]
```diff
@@ -321,8 +329,14 @@ def init_menu_profile():
...
-        endpoint='invenio_userprofiles.profile',
-        visible_when=lambda: current_user.is_authenticated,
+        endpoint='users.profile',
+        endpoint_arguments_constructor=lambda: dict(
+            viewcode=request.view_args.get(
+                'viewcode', current_app.config.get(
+                    'RERO_ILS_SEARCH_GLOBAL_VIEW_CODE'))),
+        visible_when=lambda: not current_app.config.get(
+            'RERO_PUBLIC_USERPROFILES_READONLY', False) and
+        current_user.is_authenticated,
```
Comment (zannkukai) [inline rero_ils/theme/menus.py:339]: This lambda function is used three times. You should create a function and use this function as `visible_when` argument.

**Review by zannkukai — APPROVED** [review]

**Review by BadrAly — APPROVED** [review]

**Review by pronguen — APPROVED** [review]

**Review by PascalRepond — APPROVED** [review]

File: pyproject.toml [hunk 9]
```diff
@@ -63,7 +63,7 @@ redisbeat = "*"
 jsonpickle = ">=1.4.1"
 ciso8601 = "*"
 # TODO: to be removed when the thumbnail will be refactored
-invenio-userprofiles = {git = "https://github.com/rero/invenio-userprofiles.git", rev = "v1.2.1-rero1.0"}
+invenio-userprofiles = {git = "https://github.com/rero/invenio-userprofiles.git", rev = "41d2b471cde1a93f660ba7bf0037ee3fb80b65fc"}
```
[the remark this answers is not in the document]
Comment (Garfield-fr) [reply pyproject.toml:64]: Yes, i use your PR on branch https://github.com/rero/invenio-userprofiles/tree/rero

**Pull request merged**